src/allreduce_robust.cc

```diff
@@ -600,14 +610,44 @@ AllreduceRobust::ReturnType AllreduceRobust::TryResetLinks(void) {
  * \return true if err_type is kSuccess, false otherwise
  */
 bool AllreduceRobust::CheckAndRecover(ReturnType err_type) {
+  shutdown_timeout = err_type == kSuccess;
   if (err_type == kSuccess) return true;
-  utils::Assert(err_link != NULL, "must know the error source");
-  recover_counter += 1;
 
+  _assert(err_link != NULL, "must know the error link");
+  recover_counter += 1;
+  // async launch timeout task if enable_rabit_timeout is set
+  if (rabit_timeout && !rabit_timeout_task.valid()) {
+    utils::Printf("[EXPERIMENTAL] timeout thread expires in %d second(s)\n", timeout_sec);
+    rabit_timeout_task = std::async(std::launch::async, [=]() {
+      if (rabit_debug) {
+        utils::Printf("[%d] timeout thread %ld starts\n", rank,
+                      std::this_thread::get_id());
+      }
+      int time = 0;
+      // check if rabit recovered every 100ms
+      while (time++ < 10 * timeout_sec) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        if (shutdown_timeout.load()) {
+          if (rabit_debug) {
+            utils::Printf("[%d] timeout task thread %ld exits\n",
+              rank, std::this_thread::get_id());
+          }
+          return true;
+        }
+      }
+      // print on tracker to help debuging
+      TrackerPrint("[ERROR] rank " + std::to_string(rank) + "@"+
+          host_uri + ":" +std::to_string(port) + " timeout\n");
+      _error("[%d] exit due to time out %d s\n", rank, timeout_sec);
+      return false;
+    });
+  }
   // simple way, shutdown all links
   for (size_t i = 0; i < all_links.size(); ++i) {
     if (!all_links[i].sock.BadSocket()) all_links[i].sock.Close();
   }
+  // smooth out traffic to tracker
+  std::this_thread::sleep_for(std::chrono::milliseconds(10*rank));
```

Comment:
Could you explain a little about this line?

Reply:
In our production jobs, we observed this were causing tracker failures. It all point to the fact when we have upper handreds of distributed xgboost trainers(a.k.a rabit instances). If one worker dead, the socket OOB msg will propagated to ALL trainer and starts new tcp connections to our python tracker at same time. It often overload tracker and cause entire job crash, at this point, we introduce a 10ms interval to every requests in order to mitigate this case.

Reply:
Em.. I will merge it now.  But for long term can you think of a better solution?

Reply:
Since this only happens in failure, not in normal workflow, will merge it.